src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -408,86 +434,149 @@ public MainSourceVisitor(DocumentIntermediateNode document, IntermediateNodeBuil
             //  Name=checked
             //  Prefix= checked="
             //  Suffix="
-            public override void VisitAttributeBlock(AttributeBlockChunkGenerator chunkGenerator, Block block)
+            public override SyntaxNode VisitMarkupAttributeBlock(MarkupAttributeBlockSyntax node)
             {
-                _builder.Push(new HtmlAttributeIntermediateNode()
+                var prefixTokens = MergeLiterals(
+                    node.NamePrefix?.LiteralTokens,
+                    node.Name.LiteralTokens,
+                    node.NameSuffix?.LiteralTokens,
+                    node.EqualsToken == null ? new SyntaxList<SyntaxToken>() : new SyntaxList<SyntaxToken>(node.EqualsToken),
+                    node.ValuePrefix?.LiteralTokens);
+                var prefix = (MarkupTextLiteralSyntax)SyntaxFactory.MarkupTextLiteral(prefixTokens).Green.CreateRed(node, node.NamePrefix?.Position ?? node.Name.Position);
+
+                var name = node.Name.GetContent();
+                if (name.StartsWith("data-", StringComparison.OrdinalIgnoreCase) &&
+                    !_featureFlags.EXPERIMENTAL_AllowConditionalDataDashAttributes)
                 {
-                    AttributeName = chunkGenerator.Name,
-                    Prefix = chunkGenerator.Prefix,
-                    Suffix = chunkGenerator.Suffix,
-                    Source = BuildSourceSpanFromNode(block),
-                });
+                    Visit(prefix);
+                    Visit(node.Value);
+                    Visit(node.ValueSuffix);
+                }
+                else
+                {
+                    if (node.Value != null && node.Value.ChildNodes().All(c => c is MarkupLiteralAttributeValueSyntax))
+                    {
+                        // We need to do what ConditionalAttributeCollapser used to do.
```

Comment:
Because we got rid of ConditionalAttributeCollapser, I need this logic to merge contiguous MarkupLiterals so we generate a single Intermediate node for all of it. I'll change the above comment to explain this.